### PR TITLE
add new sql auditing resources

### PR DESCRIPTION
**Pull request by yupwei68**

### CODE REVIEW

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+import (
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/services/preview/sql/mgmt/2017-03-01-preview/sql"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/satori/go.uuid"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: imports are not sorted, try run 
``` shell 
goimports -w filename
```

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+		}
+	}
+
+	var state sql.BlobAuditingPolicyState
+	switch d.Get("state").(string) {
+	case "Enabled":
+		state = sql.BlobAuditingPolicyStateEnabled
+	case "Disabled":
+		state = sql.BlobAuditingPolicyStateDisabled
+	}
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: You can use
```
state := sql.BlobAuditingPolicyState(d.Get("state").(string))
```
to replace this

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"state": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: Please add a validation function for the enum fields

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	if predictExpression, ok := d.GetOk("predicate_expression"); ok {
+		predictExpression := predictExpression.(string)
+		ExtendedServerBlobAuditingPolicyProperties.PredicateExpression = &predictExpression
+	}
+
+	parameters := sql.ExtendedServerBlobAuditingPolicy{
+		ExtendedServerBlobAuditingPolicyProperties: &ExtendedServerBlobAuditingPolicyProperties,
+	}
+
+	if _, err := client.CreateOrUpdate(ctx, resGroup, serverName, parameters); err != nil {
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: This method is an async method which returns a `future` object, please do not discard this object and use it for retrieving the result of the async operation.

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		return err
+	}
+
+	read, err := client.Get(ctx, resGroup, serverName)
+	if err != nil {
+		return err
+	}
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: This part can be replaced by handling the `future` createOrUpdate method returns.

File: azurerm/resource_arm_sql_server_blob_auditing_policies.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"state": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_server_blob_auditing_policies.go]: Please add a validation function for this field

File: azurerm/resource_arm_sql_server_blob_auditing_policies.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		}
+	}
+
+	var state sql.BlobAuditingPolicyState
+	switch d.Get("state").(string) {
+	case "Enabled":
+		state = sql.BlobAuditingPolicyStateEnabled
+	case "Disabled":
+		state = sql.BlobAuditingPolicyStateDisabled
+	}
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_sql_server_blob_auditing_policies.go]: This enum field can be replaced by 
```
state := sql.BlobAuditingPolicyState(d.Get("state").(string))
```

File: azurerm/resource_arm_storage_account.go [hunk 8]
```diff
@@ -998,6 +998,7 @@ func resourceArmStorageAccountRead(d *schema.ResourceData, meta interface{}) err
 	d.Set("secondary_access_key", "")
 
 	keys, err := client.ListKeys(ctx, resGroup, name)
+
```
Comment (ArcturusZhang) [inline azurerm/resource_arm_storage_account.go]: Redundant empty line here

File: website/docs/d/sql_extended_server_blob_auditing_policies.html.markdown [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+# Data Source: azurerm_sql_server_blob_auditing_policies
+
+Use this data source to access information about an existing SQL Azure Database Server.
+
+## Example Usage
+
+```hcl
+
+
```
Comment (ArcturusZhang) [inline website/docs/d/sql_extended_server_blob_auditing_policies.html.markdown]: Redundant empty line here

File: website/docs/d/sql_server_blob_auditing_policies.html.markdown [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+# Data Source: azurerm_sql_server_blob_auditing_policies
+
+Use this data source to access information about an existing SQL Azure Database Server Blob Auditing Policies.
+
+## Example Usage
+
+```hcl
+
+
```
Comment (ArcturusZhang) [inline website/docs/d/sql_server_blob_auditing_policies.html.markdown]: Redundant empty lines here

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"server_name": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (neil-yechenwei) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: Can we add some name validation here if azure portal also checks this schema?

Comment (yupwei68) [reply azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: No. Portal doesn't check. It's one configuration in sql server

File: azurerm/resource_arm_sql_server_blob_auditing_policies.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
+			Delete: schema.DefaultTimeout(60 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"server_name": {
```
Comment (neil-yechenwei) [inline azurerm/resource_arm_sql_server_blob_auditing_policies.go]: Can we add some name validation here if azure portal also checks this schema?

File: examples/sql_extended_server_blob_auditing_policies/main.tf [hunk 13]
```diff
@@ -0,0 +1,36 @@
+resource "azurerm_resource_group" "test" {
```
Comment (neil-yechenwei) [inline examples/sql_extended_server_blob_auditing_policies/main.tf]: Please change the alias from "test" to "example" under example folder according by hasicorp's comment. Other places with same issue also need to be updated. Thanks.

File: examples/sql_extended_server_blob_auditing_policies/main.tf [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  name     = "acctestRG"
+  location = "eastus"
```
Comment (neil-yechenwei) [inline examples/sql_extended_server_blob_auditing_policies/main.tf]: Please put these values to variables file. Other places with same issue also need to be updated. Thanks.

File: examples/sql_extended_server_blob_auditing_policies/main.tf [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+resource "azurerm_sql_server" "test" {
+  name                         = "acctestsqlserver"
```
Comment (neil-yechenwei) [inline examples/sql_extended_server_blob_auditing_policies/main.tf]: The value maybe need to updated from acctestsqlserver to example-acctestsqlserver. Other places with same issue also need to be updated.

File: website/docs/d/sql_server_blob_auditing_policies.html.markdown [hunk 16]
```diff
@@ -0,0 +1,57 @@
+---
```
Comment (neil-yechenwei) [inline website/docs/d/sql_server_blob_auditing_policies.html.markdown]: Please also add resource document for these resources.

File: azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+			"state": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+					v := val.(string)
+					if !(v == "Enabled" || v == "Disabled") {
+						errs = append(errs, fmt.Errorf("%q can only be 'Enabled' or 'Disabled' ", key))
+					}
+					return
+				},
```
Comment (njuCZ) [inline azurerm/resource_arm_sql_extended_server_blob_auditing_policies.go]: I think using
```
validation.StringInSlice([]string{
	"Enabled",
        "Disabled",
}, false),
```
to validate is more concise and elegant

**Comment by ArcturusZhang** [comment]
You may also need to update your master branch and make some empty change on your branch to update the file diffs. Now the file diff shows a lot of irrelavant things due to the master branch of the fork is out of date.

**Pull request closed without merging**